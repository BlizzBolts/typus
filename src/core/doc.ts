import ts from "typescript";
import type { SourceFile } from "./sourceFile";

export interface SerialziedDoc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: SerialziedDoc[];
  parameters?: SerialziedDoc[];
  children?: SerialziedDoc[];
  symbol?: ts.Symbol;
}

export interface DocPayload {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: Doc[];
  parameters?: Doc[];
  children?: Doc[];
  symbol?: ts.Symbol;
}

export class Doc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: Doc[];
  parameters?: Doc[];
  children?: Doc[];
  symbol?: ts.Symbol;

  root?: SourceFile | null;

  constructor(param: DocPayload) {
    this.name = param.name || "";
    this.documentation = param.documentation || "";
    this.type = param.type || "";
    this.tags = param.tags || [];
    this.required = param.required || false;
    this.members = param.members || [];
    this.parameters = param.parameters || [];
    this.children = param.children || [];
    this.symbol = param.symbol || (null as unknown as ts.Symbol);

    this.root = null;
  }

  public setRoot(root: SourceFile) {
    this.root = root;
  }

  static serialize = (curr: Doc): SerialziedDoc => {
    return {
      name: curr.name,
      documentation: curr.documentation,
      type: curr.type,
      tags: curr.tags,
      required: curr.required,
      members: curr.members?.map(Doc.serialize),
      parameters: curr.parameters?.map(Doc.serialize),
      children: curr.children?.map(Doc.serialize),
    };
  };
}
