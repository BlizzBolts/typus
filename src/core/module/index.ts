export { ModuleGraph } from "./graph";
export { Module } from "./module";
export type { BaseModule } from "./types";
