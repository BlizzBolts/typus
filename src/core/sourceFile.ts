import { Doc, SerialziedDoc } from "./doc";

export type FilePath = string;

interface SourceFilePayload {
  filePath: FilePath;
  docs?: Doc[];
}

export interface SerialzieSourceFile {
  filePath: FilePath;
  docs: SerialziedDoc[];
}

export class SourceFile {
  private filePath: FilePath;

  private docs?: Doc[];

  constructor(payload: SourceFilePayload) {
    const { filePath, docs = [] } = payload;
    this.filePath = filePath;
    this.docs = docs;
  }

  public setFilePath(filePath: FilePath): this {
    this.filePath = filePath;
    return this;
  }

  public getFilePath(): FilePath {
    return this.filePath;
  }

  public addDoc(doc: Doc) {
    this.docs!.push(doc);
    return this;
  }

  public static serialize(curr: SourceFile) {
    return {
      filePath: curr.filePath,
      docs: curr.docs?.map(Doc.serialize),
    };
  }
}
