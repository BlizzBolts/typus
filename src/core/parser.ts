import ts from "typescript";
import { generateModule } from "../core/module/utils";
import { ModuleGraph, Module } from "./module";

export const defaultCompilerOptions = {
  jsx: ts.JsxEmit.React,
  module: ts.ModuleKind.CommonJS,
  target: ts.ScriptTarget.Latest,
};

export class Parser {
  private compilerOptions: ts.CompilerOptions = defaultCompilerOptions;
  private filePaths: string[] = [];
  private program!: ts.Program;
  private typeChecker!: ts.TypeChecker;
  private moduleGraph: ModuleGraph = new ModuleGraph();

  constructor(
    filePaths?: string | string[],
    compilerOptions: Partial<ts.CompilerOptions> = defaultCompilerOptions
  ) {
    if (filePaths) {
      this.setup(filePaths, compilerOptions);
    }
  }

  static of(
    filePaths?: string | string[],
    compilerOptions: Partial<ts.CompilerOptions> = defaultCompilerOptions
  ) {
    return new Parser(filePaths, compilerOptions);
  }

  getTypeChecker() {
    return this.typeChecker;
  }

  getProgram() {
    return this.program;
  }

  getModuleGraph() {
    return this.moduleGraph;
  }

  setup(
    filePaths: string | string[],
    compilerOptions: Partial<ts.CompilerOptions> = defaultCompilerOptions
  ): this {
    this.compilerOptions = Object.assign(
      defaultCompilerOptions,
      compilerOptions
    );
    this.filePaths = Array.isArray(filePaths) ? filePaths : [filePaths];
    this.program = ts.createProgram(this.filePaths, this.compilerOptions);
    this.typeChecker = this.program.getTypeChecker();

    return this;
  }

  parse(): this {
    const validSourceFiles = this.program.getSourceFiles().filter((o) => {
      return !o.isDeclarationFile;
    });

    const moduleGraphs = validSourceFiles
      .map((o: ts.SourceFile) => {
        const result: Module[] = [];
        o.forEachChild((n: ts.Node): void => {
          const module = this.traverse(n);
          if (module) {
            result.push(module);
          }
        });
        return result;
      })
      .filter(Boolean) as ModuleGraph[][];

    this.moduleGraph.children = moduleGraphs;
    return this;
  }

  parseInterfaceDeclaration(node: ts.InterfaceDeclaration): Module {
    const type = this.typeChecker.getTypeAtLocation(node.name);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    const module = generateModule(node, type, this.typeChecker);
    this.moduleGraph.setCacheModule(type, module);
    if (Array.isArray(node.members)) {
      module.members = node.members
        .map((memberNode) => {
          return this.traverse(memberNode);
        })
        .filter(Boolean) as ModuleGraph[];
    }
    return module;
  }

  parseTypeAliasDeclaration(node: ts.TypeAliasDeclaration): Module {
    const type = this.typeChecker.getTypeAtLocation(node.name);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    if (ts.isTypeAliasDeclaration(node)) {
      console.log(1);
    }
    const module = generateModule(node, type, this.typeChecker);
    this.moduleGraph.setCacheModule(type, module);
    return module;
  }

  parseMethodSignature(node: ts.MethodSignature): Module {
    const type = this.typeChecker.getTypeAtLocation(node);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    const module = generateModule(node, type, this.typeChecker);

    if (Array.isArray(node.parameters)) {
      module.parameters = node.parameters
        .map((parameter) => this.traverse(parameter))
        .filter(Boolean) as ModuleGraph[];
    }
    return module;
  }

  parseParameter = (node: ts.ParameterDeclaration): Module => {
    const type = this.typeChecker.getTypeAtLocation(node);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    const module = generateModule(node.name, type, this.typeChecker);
    return module;
  };

  parsePropertySignature = (node: ts.PropertySignature) => {
    const type = this.typeChecker.getTypeAtLocation(node.name!);
    const cacheModule = this.moduleGraph.getCacheBy(type);
    if (cacheModule) {
      return cacheModule;
    }
    const module = generateModule(node.name, type, this.typeChecker);
    if (node.type && ts.isFunctionTypeNode(node.type)) {
      module.parameters = node.type.parameters
        .map((parameter) => this.traverse(parameter))
        .filter(Boolean) as ModuleGraph[];
    }
    return module;

    // if (node.type && ts.isTypeReferenceNode(node.type)) {
    //   node.type.parameters.forEach((parameter) =>
    //     this.traverse(parameter, doc)
    //   );
    // }
  };

  // parseFunction(node: ts.FunctionDeclaration, parent: Doc) {
  //   const fnType = this.typeChecker.getTypeAtLocation(node);
  //   const Doc = generateModule(node, fnType.getSymbol()!, this.typeChecker);
  //   parent.children?.push(Doc);
  //   this.cache.set(fnType, Doc);

  //   if (Array.isArray(node.parameters)) {
  //     node.parameters.forEach((parameter) => this.traverse(parameter, Doc));
  //   }
  // }

  // parseInterfaceMember = (node: ts.TypeElement, parent: Doc) => {
  //   const memberType = this.typeChecker.getTypeAtLocation(node!.name!);

  //   if (this.cache.has(memberType)) {
  //     parent.members?.push(this.cache.get(memberType)!);
  //   } else {
  //     const symbol = this.typeChecker.getSymbolAtLocation(node.name!);
  //     parent.members?.push(generateModule(node, symbol!, this.typeChecker));
  //   }
  // };

  traverse(node: ts.Node): Module | null {
    // if (this.cache.has(this.typeChecker.getTypeAtLocation(node))) {
    // }

    if (ts.isInterfaceDeclaration(node)) {
      console.log("isInterfaceDeclaration!");
      return this.parseInterfaceDeclaration(node);
    }

    if (ts.isMethodSignature(node)) {
      console.log("isMethodSignature!");
      return this.parseMethodSignature(node);
    }

    if (ts.isPropertySignature(node)) {
      console.log("isPropertySignature!");
      return this.parsePropertySignature(node);
    }

    if (ts.isParameter(node)) {
      console.log("isParameter!");
      return this.parseParameter(node);
    }

    // if (ts.isVariableDeclaration(node)) {
    //   console.log("isVariableDeclaration!");
    //   this.parseIndexSignatureDeclaration(node, parent)
    //   return;
    // }

    // if (ts.isFunctionDeclaration(node)) {
    //   console.log("isFunctionDeclaration!");
    //   this.parseFunction(node, parent);
    //   return;
    // }

    // if (ts.isClassDeclaration(node)) {
    //   console.log("isClassDeclaration");
    //   return;
    // }

    if (ts.isTypeAliasDeclaration(node)) {
      console.log("isTypeAliasDeclaration");
      return this.parseTypeAliasDeclaration(node);
    }

    // if (ts.isEnumDeclaration(node)) {
    //   console.log("isEnumDeclaration");
    //   return;
    // }
    // if (ts.isModuleDeclaration(node)) {
    //   console.log("isModuleDeclaration");
    //   return;
    // }

    // if (ts.isTypeElement(node)) {
    //   console.log("isTypeElement!");
    //   this.parseInterfaceMember(node, parent);
    //   return;
    // }

    return null;
  }
}
