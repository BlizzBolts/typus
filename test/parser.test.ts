import { it, expect, suite } from "vitest";
import path from "path";
import { Parser } from "../src/core/parser";

suite("core-parser", () => {
  // it("should be right cache", () => {
  //   const moduleGraphRoot = Parser.of()
  //     .setup(path.resolve(__dirname, "./manifest/parser.ts"))
  //     .parse();
  //   expect(moduleGraphRoot.toString()).toMatchSnapshot();
  // });

  it("should reuse module graph when types use same reference.", () => {
    const parser = Parser.of()
      .setup(path.resolve(__dirname, "./manifest/reused-interface"))
      .parse();

    expect(parser.getModuleGraph().hitCacheTimes).equal(1);
  });

  it("shouldn't reuse any module graph when types are totally 1 diff", () => {
    const parser = Parser.of()
      .setup(path.resolve(__dirname, "./manifest/not-reused"))
      .parse();
    expect(parser.getModuleGraph().hitCacheTimes).equal(0);
  });
});
