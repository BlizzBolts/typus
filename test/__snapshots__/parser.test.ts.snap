// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`core-parser > should be right cache 1`] = `
{
  "children": [
    [
      {
        "children": [],
        "documentation": "",
        "members": [
          {
            "children": [],
            "documentation": "上报日志",
            "members": [],
            "name": "log",
            "parameters": [
              {
                "children": [],
                "documentation": "",
                "members": [],
                "name": "payload",
                "parameters": [],
                "required": false,
                "tags": [],
                "type": "{ content: string; v: number; }",
              },
            ],
            "required": false,
            "tags": [
              {
                "name": "date",
                "text": "7/4/2023 - 4:43:53 PM",
              },
              {
                "name": "type",
                "text": "{(payload: { content: string, v: number }) => Promise<void>}",
              },
            ],
            "type": "(payload: { content: string; v: number; }) => Promise<void>",
          },
        ],
        "name": "ActionMeta",
        "parameters": [],
        "required": false,
        "tags": [],
        "type": "any",
      },
    ],
  ],
}
`;
